src/artemis/document.cljc

```diff
@@ -1,66 +1,123 @@
 
-(s/fdef doc?
-        :args (s/cat :x any?)
-        :ret  boolean?)
+(s/fdef operation
+        :args (s/alt
+                :arity-1 (s/cat :document ::document)
+                :arity-2 (s/cat :document  ::document
+                                :variables map?)
+                :arity-3 (s/cat :document          ::document
+                                :variables         map?
+                                :inline-fragments? boolean?))
+        :ret  ::operation)
 
-(defn doc?
-  "Returns `true` if `x` is a GraphQL document."
+(defn operation
+  "Given a document and some variables, return a map that describes the
+  operation to execute under the :graphql key and a function to neatly unpack
+  the data that results from that operation under the :unpack key. Unpacking
+  reifies operations that were namespaced as a result of applying an operation
+  mapping."
   {:added "0.1.0"}
-  [x]
-  (instance? Document x))
-
-(s/def ::document doc?)
-(s/def ::source string?)
-(s/def ::ast vector?)
+  ([document]
+   (operation document {} false))
+  ([document variables]
+   (operation document variables false))
+  ([document variables inline-fragments?]
+   (if (single-operation? document)
+     (let [op (-> document :operation-definitions first :operation-type)]
+       ((get-in (gql/query-map document {:inline-fragments inline-fragments?})
+                [(keyword (:type op))
+                 (keyword (->kebab-case (:name op)))])
+        variables))
+     (let [op-map (::operation-mapping (meta document) {})]
+       ((gql/composed-query document op-map) variables)))))
 
-(s/fdef source
-        :args (s/cat :document ::document)
-        :ret  (s/or :nil    nil?
-                    :source ::source))
+(s/fdef compose
+        :args (s/cat :doc  ::document
+                     :docs (s/* ::document))
+        :ret  ::document)
 
-(defn source
-  "Returns the string source for a GraphQL document."
+(defn compose
+  "Compose multiple parsed documents together to create a single document."
   {:added "0.1.0"}
-  [document]
-  (:source document))
+  [doc & docs]
+  (apply merge-with into doc docs))
 
-(s/fdef ast
-        :args (s/cat :document ::document)
-        :ret  (s/or :nil    nil?
-                    :source ::ast))
+(s/fdef with-mapping
+        :args (s/cat :doc     ::document
+                     :mapping map?)
+        :ret  ::document)
 
-(defn ast
-  "Returns the alumbra-parsed AST for a GraphQL document."
+(defn with-mapping
+  "Attaches an operation mapping to a document as meta data. An operation
+  mapping maps a key to any named GraphQL operation within the document.
+  Different keys can map to the same operation. When running an operation from
+  the document, the operation mapping will be used to resolve multiple sets of
+  variables based on the mapped key."
   {:added "0.1.0"}
-  [document]
-  (:ast document))
+  [doc mapping]
+  (vary-meta doc assoc ::operation-mapping mapping))
+
+#?(:clj (do
```

Comment:
Indentation is weird here - what is the `do` for?

Reply:
When you're doing this reader macro stuff, you can only give it a single expression, so if you want to do a bunch of stuff (like define things) you have to wrap them in a do:

Check out the section right Host Interop here https://clojure.org/guides/reader_conditionals
As for indentation, that's just the style I've seen most often, bringing that `(do` part up there 
and then pretending it doesn' exist. I'm happy to follow the "rules" a bit more with keeping the expressions in line with the `do`

Reply:
That's cool, I'm just unfamiliar with this syntax/style - thanks for the info.